pyproject.toml

```
@@ -1,15 +1,82 @@
 [build-system]
-build-backend = "setuptools.build_meta"
+build-backend = "mesonpy"
 requires = [
-    "setuptools>=42",
-    "setuptools_scm[toml]>=7.0.0",
+    "meson-python",
     "pybind11>=2.8.0",
     "wheel",
+    "versioningit",
 ]
 
-[tool.setuptools_scm]
-version_scheme = "no-guess-dev"
-version_file = "pyamg/version.py"
+[project]
+name = "pyamg"
+dynamic = ["version"]
+authors = [
+    {name = "Nathan Bell"},
+    {name = "Luke Olson", email = "luke.olson@gmail.com"},
+    {name = "Jacob Schroder"},
+    {name = "Ben Southworth"},
+]
+maintainers = [
+    {name = "Luke Olson", email = "luke.olson@gmail.com"},
+    {name = "Jacob Schroder"},
+    {name = "Ben Southworth"},
+]
+description = "PyAMG: Algebraic Multigrid Solvers in Python"
+readme = "README.md"
+license = {text = "MIT License"}
+keywords = ["algebraic multigrid", "AMG", "sparse matrix", "preconditioning"]
+classifiers = [
+    "Development Status :: 5 - Production/Stable",
+    "Intended Audience :: Developers",
+    "Intended Audience :: Education",
+    "Intended Audience :: Science/Research",
+    "License :: OSI Approved :: MIT License",
+    "Natural Language :: English",
+    "Operating System :: OS Independent",
+    "Programming Language :: C++",
+    "Programming Language :: Python",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
+    "Topic :: Education",
+    "Topic :: Scientific/Engineering",
+    "Topic :: Scientific/Engineering :: Mathematics",
+    "Topic :: Software Development :: Libraries :: Python Modules",
+]
+dependencies = [
+    "numpy",
+    "scipy>=1.11.0"
+]
+
+[project.optional-dependencies]
+test = [
+    "pytest",
+]
+dev = [
+    "meson-python",
+    "meson",
+    "ninja",
+    "pybind11",
+    "ruff",
+    "pytest",
+]
```

Totally up to you how to break these up. This is subjective ofc

Just as another perspective, this is what we did with [our optionals in Zarr]( https://github.com/zarr-developers/zarr-python/blob/df4c25f70c8a1e2b43214d7f26e80d34df502e7e/pyproject.toml#L57-L81 )

Maybe one consideration to guide the grouping is to think about when they are used. For example, would users wanting to run a linter or fix style want to build from source? If test dependencies are in one option, would it make sense to include in others or would users simply mix the two groups to get what they want

Just to illustrate this a bit, one could consider

```suggestion
[project.optional-dependencies]
build = [
    "meson-python",
    "meson",
    "ninja",
    "pybind11",
]
style = [
    "ruff",
]
test = [
    "pytest",
]
```

Then if one wanted all of them, they could do `pip install .[build,style,test]`

Though again this is all subjective and simply depends on the kind of experience you and other devs want. Ultimately it matters more how you are using these to decide how to shape them

---

thanks for the suggestion; this is great